### PR TITLE
IGNITE-20662 Sql. Added a microbenchmark to measure script execution performance

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-20662

Added new single statement benchmarks to measure overhead of script execution:
```
InsertBenchmark.sqlInsertScript
InsertBenchmark.jdbcInsertScript

SelectBenchmark.sqlGetScript
SelectBenchmark.jdbcGetScript
```

Added `SqlMultiStatementBenchmark`, which compares executing statements sequentially with executing the same statements as a single script. The number of statements is set by the benchmark parameter.

The following scenarios have been added to it:

INSERT
```
INSERT INTO TABLE1 VALUES(...);
INSERT INTO TABLE2 VALUES(...);
...
INSERT INTO TABLEN VALUES(...);
```

COUNT
```
SELECT COUNT(*) FROM TABLE1;
SELECT COUNT(*) FROM TABLE2;
...
```

SELECT MULTIPLE TABLES
```
SELECT * FROM TABLE1 where key = ?;
SELECT * FROM TABLE2 where key = ?;
...
```

SELECT MULTIPLE KEYS SAME TABLE
```
SELECT * FROM TABLE1 where key = Key1;
SELECT * FROM TABLE1 where key = Key2;
...
SELECT * FROM TABLE1 where key = KeyN;
```

Results with changes from PR #2965 (`SELECT` parallelism)

```
Benchmark                                              (statementsCount)  Mode  Cnt    Score   Error  Units
SqlMultiStatementBenchmark.count                                       1  avgt   10   11.880 ± 2.162  ms/op
SqlMultiStatementBenchmark.countScript                                 1  avgt   10   12.351 ± 1.077  ms/op
SqlMultiStatementBenchmark.count                                       2  avgt   10   26.123 ± 4.019  ms/op
SqlMultiStatementBenchmark.countScript                                 2  avgt   10   13.931 ± 2.237  ms/op
SqlMultiStatementBenchmark.count                                       4  avgt   10   54.906 ± 4.226  ms/op
SqlMultiStatementBenchmark.countScript                                 4  avgt   10   14.250 ± 1.315  ms/op
SqlMultiStatementBenchmark.count                                       8  avgt   10  110.833 ± 9.095  ms/op
SqlMultiStatementBenchmark.countScript                                 8  avgt   10   28.572 ± 5.199  ms/op

SqlMultiStatementBenchmark.insert                                      1  avgt   10    1.173 ± 0.091  ms/op
SqlMultiStatementBenchmark.insertScript                                1  avgt   10    1.405 ± 0.142  ms/op
SqlMultiStatementBenchmark.insert                                      2  avgt   10    2.394 ± 0.166  ms/op
SqlMultiStatementBenchmark.insertScript                                2  avgt   10    2.756 ± 0.135  ms/op
SqlMultiStatementBenchmark.insert                                      4  avgt   10    4.984 ± 0.313  ms/op
SqlMultiStatementBenchmark.insertScript                                4  avgt   10    5.362 ± 0.444  ms/op
SqlMultiStatementBenchmark.insert                                      8  avgt   10   10.573 ± 0.503  ms/op
SqlMultiStatementBenchmark.insertScript                                8  avgt   10   10.729 ± 1.190  ms/op

SqlMultiStatementBenchmark.selectMultipleKeys                          1  avgt   10    0.753 ± 0.096  ms/op
SqlMultiStatementBenchmark.selectMultipleKeysScript                    1  avgt   10    0.870 ± 0.134  ms/op
SqlMultiStatementBenchmark.selectMultipleKeys                          2  avgt   10    1.517 ± 0.104  ms/op
SqlMultiStatementBenchmark.selectMultipleKeysScript                    2  avgt   10    1.226 ± 0.094  ms/op
SqlMultiStatementBenchmark.selectMultipleKeys                          4  avgt   10    3.157 ± 0.332  ms/op
SqlMultiStatementBenchmark.selectMultipleKeysScript                    4  avgt   10    1.618 ± 0.176  ms/op
SqlMultiStatementBenchmark.selectMultipleKeys                          8  avgt   10    7.021 ± 0.744  ms/op
SqlMultiStatementBenchmark.selectMultipleKeysScript                    8  avgt   10    2.419 ± 0.425  ms/op

SqlMultiStatementBenchmark.selectMultipleTables                        1  avgt   10    0.760 ± 0.100  ms/op
SqlMultiStatementBenchmark.selectMultipleTablesScript                  1  avgt   10    0.878 ± 0.085  ms/op
SqlMultiStatementBenchmark.selectMultipleTables                        2  avgt   10    1.565 ± 0.170  ms/op
SqlMultiStatementBenchmark.selectMultipleTablesScript                  2  avgt   10    1.224 ± 0.069  ms/op
SqlMultiStatementBenchmark.selectMultipleTables                        4  avgt   10    3.267 ± 0.353  ms/op
SqlMultiStatementBenchmark.selectMultipleTablesScript                  4  avgt   10    1.611 ± 0.193  ms/op
SqlMultiStatementBenchmark.selectMultipleTables                        8  avgt   10    7.209 ± 0.681  ms/op
SqlMultiStatementBenchmark.selectMultipleTablesScript                  8  avgt   10    2.554 ± 0.434  ms/op
```

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SqlMultiStatementBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+        }
+
+        private int id = 0;
+
+        void executeQuery(Blackhole bh) {
+            int id0 = id++;
+
+            for (int i = 0; i < statementsCount; i++) {
+                try (ResultSet<?> rs = session.execute(null, parameters.statements.get(i), id0)) {
+                    // NO-OP
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SqlMultiStatementBenchmark.java:181]: bh.consume(rs);

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SelectBenchmark.java [hunk 2]
```diff
@@ -122,6 +138,19 @@ public void jdbcGet(JdbcState state) throws SQLException {
...
+    /**
+     * Benchmark for JDBC script get.
+     */
+    @Benchmark
+    public void jdbcGetScript(JdbcState state) throws SQLException {
+        state.stmt.setInt(1, random.nextInt(TABLE_SIZE));
+        state.stmt.execute();
+
+        try (ResultSet r = state.stmt.getResultSet()) {
+            r.next();
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SelectBenchmark.java:150]: lets add Blackhole here too ?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SqlMultiStatementBenchmark.java (hunk 1 above)
Comment (xtern) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SqlMultiStatementBenchmark.java:181]: Done, thanks

**Review by zstan — APPROVED** [review]

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SelectBenchmark.java [hunk 3]
```diff
@@ -172,7 +204,40 @@ private org.apache.ignite.sql.ResultSet<SqlRow> sql(String sql, Object... args)
     }
 
     /**
-     * Benchmark state for {@link #sqlThinGet(SqlThinState)}.
+     * Benchmark state for {@link #sqlGet(SqlState, Blackhole)}.
+     *
+     * <p>Holds {@link Session}.
+     */
+    @State(Scope.Benchmark)
+    public static class SqlScriptState {
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SelectBenchmark.java:212]: to make fair comparison, we need one more state: the one that uses internal API as well, but instead executes statements with `querySingleAsync`

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SqlMultiStatementBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+/**
+ * Benchmark measures the time for sequential execution of different sets of statements and the time
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SqlMultiStatementBenchmark.java:64]: we need to compare different modes of the same API. If at the moment we can't run scripts with any API but internal, then we need to run benchmarks only using internal API. Later we will add the similar set of benchmarks for JDBC and java API

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SelectBenchmark.java (hunk 3 above)
Comment (xtern) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/SelectBenchmark.java]: Thanks. Added a separate `sqlGetInternal` benchmark that uses the internal API.

**Pull request merged**